In [15]:
print("NEWTON METHOD AND BISECTION METHOD COMPARISON")

NEWTON METHOD AND BISECTION METHOD COMPARISON


In [22]:
%%timeit -r 4 -n 10000
#newton rhapshon
import numpy as np

f = lambda x: x**4 - 2
f_prime = lambda x: 4*x**3

def newton(f, df, x0, tol):
    # output is an estimation of the root of f using the Newton Raphson method
    if abs(f(x0)) < tol:
        return x0
    else:
        #recursive call
        return newton(f, df, x0 - f(x0)/df(x0), tol)



1.38 µs ± 27.6 ns per loop (mean ± std. dev. of 4 runs, 10000 loops each)


$$
\ f(x)=x^{4}-2,  f'(x)=4x^{3}
$$

In [31]:
%%timeit -r 4 -n 10000
import numpy as np

def bisection(f, a, b, tol): 
    # approximates a root of f bounded by a and b to within tolerance 
    # absolute f(m) < tol with m the midpoint 
    
    # check if a and b bound a root
    if np.sign(f(a)) == np.sign(f(b)):
        return"The scalars a and b do not bound a root"
        
    # get midpoint
    m = (a + b)/2
    
    if np.abs(f(m)) < tol:
        #report m as root
        return m
    elif np.sign(f(a)) == np.sign(f(m)):
        # when m is an improvement on a, make a recursive call with a = m
        return my_bisection(f, m, b, tol)
    elif np.sign(f(b)) == np.sign(f(m)):
        # when m is an improvement on b, make recursive call with b = m
        return my_bisection(f, a, m, tol)
    
f = lambda x: x**4 - 2


1.06 µs ± 29.8 ns per loop (mean ± std. dev. of 4 runs, 10000 loops each)


$$
\ f(x)= x^{4}-2
$$

$$Newton\ raphson\ method\ is\ faster\ than\ Bisection\ method\ with\ a\ speed\  of\  1.06\mu s\  compared\\  to\ bisection\ methods\ 1.38\mu s
$$

$$ IMPLEMENTATION\ OF\ ACTUAL\ CODE\ IS\ BELOW\ \downarrow
$$

In [29]:
#newton rhapshon
import numpy as np
x0=3
f = lambda x: x**4 - 2
f_prime = lambda x: 4*x**3
tol=1e-6

def newton(f, df, x0, tol):
    if abs(f(x0)) < tol:
      print(x0)
    else:
        return newton(f, df, x0 - f(x0)/df(x0), tol)

newton(f,f_prime,x0,tol)

1.1892071150571961


In [30]:
import numpy as np

def bisection(f, a, b, tol):
    
    # check if a and b have a root
    if np.sign(f(a)) == np.sign(f(b)):
        return "The scalars a and b do not bound a root"
    # get midpoint
    m = (a + b)/2
    
    if np.abs(f(m)) < tol:
        #  print m as root
        print (m)
    elif np.sign(f(a)) == np.sign(f(m)): 
        # Make recursive call with a = m
        return bisection(f, m, b, tol)
    elif np.sign(f(b)) == np.sign(f(m)):
        # Make recursive call with b = m
        return bisection(f, a, m, tol)
    
f = lambda x: x**4 - 2
#tolerance
tol=1e-6
#call function set bounds within which function will operate
bisection(f,1,10,tol)

1.1892070472240448


$$ Both\ methods\ return\ same\ values\ for\ root\ with\ minor\ differences\ in\ the\ lower\ decimal\ places $$

$$ \underline{CONCLUSION} \\
EVEN\ THOUGH\ BOTH\ METHODS\ CAN\ BE\ USED\ TO\ FIND\ ROOTS\ THE\ NEWTON\ RAPHSON\ METHOD\ IS\ FASTER\
$$